Check server state when bring nodes back

SidW opened the pull request:
Fixed the issue when starting up Voldemort in offilne state, it actually
goes online and keeps listening client requests.

arunthirupathi commented on src/java/voldemort/server/VoldemortServer.java:
```diff
@@ -382,7 +382,15 @@ protected void startInner() throws VoldemortException {
+        else if (getMetadataStore().getServerStateUnlocked() == MetadataStore.VoldemortState.OFFLINE_SERVER)
+            goOnline = false;
+        else
+            throw new VoldemortException("Invalid Server state. Neither online nor offline found");
```
The server has 3 states, online, offline and rebalancing. This makes the server not to start when that is in the Rebalancing mode. I would like to have only 2 cases here, offline and everything else.

arunthirupathi commented on test/unit/voldemort/client/OfflineStateTest.java:
```diff
@@ -262,4 +268,23 @@ public void testStateTransitions() throws InterruptedException {
+                    ServerTestUtils.createServerConfigWithDefs(useNio, 0, parentDirPath, cluster, storeDefs, new Properties());
+            VoldemortServer newServer = ServerTestUtils.startVoldemortServer(socketStoreFactory, config, cluster);
+            fail("should have thrown an connection exception");
+        }catch (VoldemortException e){
```
I would just handle Catch UnreachableStoreException and also check that the inner exception is ConnectException.class.

arunthirupathi commented on test/unit/voldemort/client/OfflineStateTest.java:
```diff
@@ -262,4 +268,23 @@ public void testStateTransitions() throws InterruptedException {
+        }catch (VoldemortException e){
+            assertEquals(UnreachableStoreException.class, e.getClass());
+        }
+    }
```
I would also like to add that the server that was brought up in offline mode can go to online mode. by calling the appropriate methods.

arunthirupathi commented on test/unit/voldemort/client/OfflineStateTest.java:
```diff
@@ -262,4 +268,23 @@ public void testStateTransitions() throws InterruptedException {
         assertTrue(testSlopStreaming());
     }
+
+    @Test
```
I would also like to add one more test, where a server stopped in online mode, can be brought up in offline mode. The steps that are done in the test, need to be documented and given to SREs.

arunthirupathi commented on src/java/voldemort/server/VoldemortServer.java:
```diff
@@ -386,10 +386,8 @@ protected void startInner() throws VoldemortException {
         boolean goOnline;
         if (getMetadataStore().getServerStateUnlocked() == MetadataStore.VoldemortState.NORMAL_SERVER)
```
Rebalance Server should still come as online. This would make the rebalance server to go offline.

arunthirupathi commented:
The test looks good now. If you can make minor tweak to the original logic and squash your commit, I will merge the commit.

SidW commented:
Thanks Arun! I should have been more careful. I have modified that and
rebased my commits. Ready to be merged😀

On Wed, Mar 23, 2016 at 1:24 PM, ARUNACHALAM THIRUPATHI <
notifications@github.com> wrote:

> The test looks good now. If you can make minor tweak to the original logic
> and squash your commit, I will merge the commit.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/voldemort/voldemort/pull/388#issuecomment-200528978>
>

arunthirupathi commented:
Rebased and merged  https://github.com/voldemort/voldemort/commit/af3d605f9b953f1caa0bd36cde3f1e3a0635da3e
